### ISSUE TITLE
Refetch triggered when an unrelated query is made

### ISSUE THREAD

**Issue by lwasyl** [apollographql/apollo-android#2798]
(Apollo 2.4.6)
Not sure if this is a bug or a question 🙂

Given the following schema:
```graphql
type Query {
  foo: Foo!
  bar: Bar!
}

type Foo {
  fff: String!
}

type Bar {
  bbb: String!
}
```
and two unrelated queries:
```graphql
query Foo {
  foo {
    fff
  }
}

query Bar {
  bar {
    bbb
  }
}
```

the following test (mock web server omitted, but it enqueues the responses and fails immediately when a response was not enqueued) will fail due API call for `Foo` query being made.
```kotlin
@Test
internal fun `disjoint query makes network request`(): Unit = runBlocking {
  val apollo = buildApollo()

  val channel = Channel<Response<FooQuery.Data>>(capacity = Channel.UNLIMITED)
  val job = launch {
    apollo.query(FooQuery()).toBuilder()
      .responseFetcher(ApolloResponseFetchers.CACHE_ONLY)
      .build()
      .watcher()
      .refetchResponseFetcher(ApolloResponseFetchers.CACHE_FIRST)
      .toFlow()
      .collect {
        channel.sendBlocking(it)
      }
  }

  channel.receive() // wait for the first, empty response

  mockWebServer.enqueue(
    """{
    "data": {
      "bar": {
        "__typename": "Bar",
        "bbb": "bbb"
      }
    }
  }"""
  )

  apollo.query(BarQuery()).toBuilder()
    .responseFetcher(ApolloResponseFetchers.NETWORK_ONLY)
    .build()
    .await()

  delay(500) // Allow some time for Apollo to trigger stuff, for some reason

  job.cancel()
}

private fun buildApollo() = ApolloClient
  .builder()
  .normalizedCache(LruNormalizedCacheFactory(EvictionPolicy.NO_EVICTION), object : CacheKeyResolver() {
    override fun fromFieldRecordSet(field: ResponseField, recordSet: Map<String, Any>): CacheKey =
      (recordSet["id"] as? String).asCacheKey()

    override fun fromFieldArguments(field: ResponseField, variables: Operation.Variables): CacheKey =
      (field.resolveArgument("id", variables) as? String).asCacheKey()

    private fun String?.asCacheKey(): CacheKey {
      if (this.isNullOrBlank()) return CacheKey.NO_KEY

      return CacheKey(this)
    }
  })
  .serverUrl(mockWebServer.mockWebServer.url("/"))
  .dispatcher(immediateExecutorService())
  .okHttpClient(OkHttpClient.Builder().dispatcher(Dispatcher(immediateExecutorService())).build())
  .build()
```

Key pieces: 
 - first query is made with `CACHE_ONLY` fetcher and a `CACHE_FIRST` refetcher. 
 - wait until the first watcher emits empty data
 - second query makes a network request
 - first query attempts to (re)fetch the data

I'm not sure if this is by design, but it's pretty surprising behavior. Why does a completely unrelated query cause `FooQuery` watcher to refetch the data? On one hand it does make some sense (there _is_ a cache miss), but on the other the data that `FooQuery` would watch didn't change. This becomes a bigger problem when we start watching several queries using `CACHE_ONLY` fetcher and a refetcher that allows network - I assume all of them will trigger a network call as soon as the cache is populated?

**Comment by martinbonnin** [comment]
I think this is all related to the fact that query Foo doesn't know its dependentKeys before a request is successfully executed. That's the whole [if (dependentKeys.isEmpty() refetch thing](https://github.com/apollographql/apollo-android/blob/621c1a192e9434a8981d5ce6825e5c05df79e454/apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloQueryWatcher.java#L40).

Thinking about it, I think we should be able to compute a first version of the dependent keys from the query alone, to restrict the false positives? In that first query above, we would start with something like `CacheKey("QUERY_ROOT/foo")`. And then update the dependentKeys once we get the response. I think that might work?

**Comment by lwasyl** [comment]
That'd make sense, but would still sometimes trigger the refetch when not needed, right? E.g. when some other query fetches `foo { somethingElse }`. I'm wondering if this is something that even needs fixing, what we want to do is to keep queries `cache_only` and only populate the cache from prefetches/network_only queries. Perhaps it'd work better for us to have a custom refetch fetcher that doesn't allow network access until the watcher has emitted at least once (which would be when another instance of this query with network_only is made)?

**Comment by martinbonnin** [comment]
> E.g. when some other query fetches foo { somethingElse }

Correct

> Perhaps it'd work better for us to have a custom refetch fetcher that doesn't allow network access until the watcher has emitted at least once (which would be when another instance of this query with network_only is made)?

Wouldn't that defeat the fact that you want your first query to be `cache_only` ? If you're open to changing that, what about making the query with `cache_first` and the refetcher `cache_only` ? That'll ensure the network is only reached once?

**Comment by lwasyl** [comment]
I realized the refetcher must be `cache_first`, because otherwise we'd end up either with stale data (or no data, if the query emits response with `null` data) or having to manually make a network request regardless (which is something to consider). The benefit of `network_only` is primarily to keep unidirectional data flow, and queries that have side effects violate that. Practically, this also helps us in UI tests, where we enqueue responses and fail if a response we didn't anticipate was made. Splitting the actions into strictly observing and fetching helps a lot it understanding the data flow in the app.

My current understanding is that `network_only` fetches and `cache_only` queries should cover most of the cases. Ideally the refetches wouldn't have to happen at all, but in an off chance the schema and queries are structured such that it's not practical, then we allow refetches as the only way to keep the data consistent.

**Comment by martinbonnin** [comment]
> I realized the refetcher must be cache_first, because otherwise we'd end up either with stale data (or no data, if the query emits response with null data) 

I am confused. If the query1 watcher is `cache_only` (both query and refetcher) and query2 is `network_only`, ultimately the watcher should receive the correct data ? Even if there are intermediate cache misses with `null` data ?

**Comment by lwasyl** [comment]
The first emission, yes. But after some data was emitted already, then when subsequent cache misses happen we'd have to either:
 - ignore intermediate cache misses with null data -> we'd have stale information
 - _lose_ the data briefly when intermediate cache miss comes -> poor UX, as we use Apollo as the single source of truth and the data we observe is propagated to the UI

So the `cache_only` refetch strategy only works when fetching the data for the first time, but later on the cache misses complicate things. Hope I'm clearing things up a bit 🙂

**Comment by martinbonnin** [comment]
Thanks for the details!

> ignore intermediate cache misses with null data -> we'd have stale information

Right! If some other `query3` that partially overlaps changes an id or adds to a list, the `query1` watcher will not see the changes.

Thinking out loud: since you seem to have a concept of `initial network query` + `subsequent incremental updates`, could you potentially wait until the `network_only` query is done before starting the watchers?

**Comment by lwasyl** [comment]
>  since you seem to have a concept of initial network query + subsequent incremental updates, could you potentially wait until the network_only query is done before starting the watchers?

Yes, you're right. My conclusion is that this a pretty specific requirement and we should be able to implement it using the existing APIs. I was thinking about a custom response fetcher (/wrapper) which would be a CACHE_ONLY until the first cache hit and CACHE_FIRST afterwards, for example.

As for computing the initial set of dependent keys from the query - it's reasonable to limit unexpected network calls. Since there's still a chance for false positives and I now see that we'll need some custom handling here anyway, it's not impacting us anymore so feel free to close the issue (unless you want to keep it open to track this specific improvement). Thanks!

**Comment by martinbonnin** [comment]
I'm not sure computing the initial set of dependent keys from the query is a good idea ultimately. In this scenario that queries a list of messages:

```
{
  messages {
    id
    content
    subject
  } 
}
```

Changing the content/subject of a message will not trigger the refetch despite this being part of the query so that can be surprising. 

Also, I learned iOS has the opposite default behaviour: if the cache is initially empty, they do not refetch on cache updates, which has another set of advantages and drawbacks. 

Maybe in the end, what we want is an API that lets the users decide whether they want to refetch or not. I'll close this issue for now and we can reopen a new one once we have identified a good use case.
